access/ms-teams/bot.go

```diff
@@ -125,52 +136,72 @@ func (b Bot) UninstallAppForUser(ctx context.Context, userIDOrEmail string) erro
 	return trace.Wrap(err)
 }
 
-// FetchUser fetches app id for user, installs app for a user if missing, fetches chat id and saves
-// everything to cache. This method is used for priming the cache. Returns trace.NotFound if a
+// FetchRecipient checks if recipient is a user or a channel, installs app for a user if missing, fetches chat id
+// and saves everything to cache. This method is used for priming the cache. Returns trace.NotFound if a
 // user was not found.
-func (b Bot) FetchUser(ctx context.Context, userIDOrEmail string) (*UserData, error) {
+func (b Bot) FetchRecipient(ctx context.Context, recipient string) (*RecipientData, error) {
 	if b.teamsApp == nil {
 		return nil, trace.Errorf("Bot is not configured, run GetTeamsApp first")
 	}
 
 	b.mu.RLock()
-	d, ok := b.users[userIDOrEmail]
+	d, ok := b.recipients[recipient]
 	b.mu.RUnlock()
 	if ok {
 		return &d, nil
 	}
 
-	userID, err := b.getUserID(ctx, userIDOrEmail)
-	if err != nil {
-		return &UserData{}, trace.Wrap(err)
-	}
-
-	var installedApp *msapi.InstalledApp
-
-	installedApp, err = b.graphClient.GetAppForUser(ctx, b.teamsApp, userID)
-	if trace.IsNotFound(err) {
-		err := b.graphClient.InstallAppForUser(ctx, userID, b.teamsApp.ID)
+	// Check if the recipient is a channel
+	channel, isChannel := b.checkChannelURL(recipient)
+	if isChannel {
+		// A team and a group are different but in MsTeams the team is associated to a group and will have the same id.
+		installedApp, err := b.graphClient.GetAppForTeam(ctx, b.teamsApp, channel.Group)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
+		d = RecipientData{
+			ID:  fmt.Sprintf("%s/%s/%s", channel.Tenant, channel.Group, channel.Name),
+			App: *installedApp,
+			Chat: msapi.Chat{
+				ID:       channel.ChatID,
+				TenantID: channel.Tenant,
+				WebURL:   channel.URL.String(),
+			},
+		}
+		// If the recipient is not a channel, it means it is a user (either email or userID)
+	} else {
+		userID, err := b.getUserID(ctx, recipient)
+		if err != nil {
+			return &RecipientData{}, trace.Wrap(err)
+		}
+
+		var installedApp *msapi.InstalledApp
 
 		installedApp, err = b.graphClient.GetAppForUser(ctx, b.teamsApp, userID)
+		if trace.IsNotFound(err) {
+			err := b.graphClient.InstallAppForUser(ctx, userID, b.teamsApp.ID)
```

Comment:
Is it OK that this could be called more than once due to cache misses? Does "already installed" manifest as an error?

Reply:
Yup, this could be an issue as discussed here: https://github.com/gravitational/teleport-plugins/pull/646#discussion_r961470167

https://github.com/gravitational/teleport-plugins/pull/646/commits/1fbe4bd0d517d176b498740c7f6999e04f981416 should handle this case